kernel/safe_typing.ml

```diff
@@ -957,15 +989,12 @@ let add_private_constant l uctx decl senv : (Constant.t * private_constants) * s
       | DefinitionEff ce ->
         let () = assert (check_constraints uctx ce.Entries.const_entry_universes) in
         Term_typing.translate_constant ~sec_univs senv.env kn (Entries.DefinitionEntry ce)
-    in
-  let dcb = match cb.const_body with
-  | Def _ as const_body -> { cb with const_body }
+      in
+  let senv, dcb = match cb.const_body with
+  | Def _ as const_body -> senv, { cb with const_body }
   | OpaqueDef _ ->
-    (* We drop the body, to save the definition of an opaque and thus its
-       hashconsing. It does not matter since this only happens inside a proof,
-       and depending of the opaque status of the latter, this proof term will be
-       either inlined or reexported. *)
-    { cb with const_body = Undef None }
+      let senv, o = push_opaque_proof senv in
+      senv, { cb with const_body = OpaqueDef o }
```

Comment:
This essentially reverts https://github.com/coq/coq/pull/12935, I guess that's why this PR fixes #13324

Reply:
Beware that while #12935 caused the observable bug, as commented in #13324 the previous state was incorrect in a different way, i.e. gladly keeping around dead subproofs. We should check what the current does exactly.

Reply:
I checked the vo file size on the bench, and it didn't seem to vary that much. Something may have happened in the meantime that solved the underlying bug?